mwdb/model/file.py

```diff
@@ -328,3 +329,210 @@ def get_by_download_token(download_token):
+            os.makedirs(upload_path, mode=0o755, exist_ok=True)
+
+        return os.path.join(upload_path, sample_sha256)
+
+    @classmethod
+    def create(
+        cls,
+        file_name,
+        file_stream,
+        main_obj_dhash,
+    ):
+        file_stream.seek(0, os.SEEK_END)
+        file_size = file_stream.tell()
+        if file_size == 0:
+            raise EmptyFileError
+
+        sha256 = calc_hash(file_stream, hashlib.sha256(), lambda h: h.hexdigest())
+
+        new_related_file = (
+            db.session.query(RelatedFile).filter(RelatedFile.sha256 == sha256).first()
+        )
+        main_obj = (
+            db.session.query(Object).filter(Object.dhash == main_obj_dhash).first()
+        )
+
+        # If file already exists
+        if new_related_file is not None:
+            raise FileExistsError("Related file with this sha256 already exists")
+
+        # If related file doesn't exist
+        if main_obj is None:
+            raise ValueError(
+                "There is no object with this sha256 or you don't have access"
+            )
+
+        new_related_file = RelatedFile(
+            object_id=main_obj.id,
+            file_name=secure_filename(file_name),
+            file_size=file_size,
+            sha256=sha256,
+        )
+
+        file_stream.seek(0, os.SEEK_SET)
+        if app_config.mwdb.storage_provider == StorageProviderType.S3:
+            get_s3_client(
+                app_config.mwdb.s3_storage_endpoint,
+                app_config.mwdb.s3_storage_access_key,
+                app_config.mwdb.s3_storage_secret_key,
+                app_config.mwdb.s3_storage_region_name,
+                app_config.mwdb.s3_storage_secure,
+                app_config.mwdb.s3_storage_iam_auth,
+            ).put_object(
+                Bucket=app_config.mwdb.s3_storage_bucket_name,
+                Key=new_related_file._calculate_path(),
+                Body=file_stream,
+            )
+        elif app_config.mwdb.storage_provider == StorageProviderType.DISK:
+            with open(new_related_file._calculate_path(), "wb") as f:
+                shutil.copyfileobj(file_stream, f)
+        else:
+            raise RuntimeError(
+                f"StorageProvider {app_config.mwdb.storage_provider} "
+                f"is not supported"
+            )
+        db.session.add(new_related_file)
+        db.session.commit()
+
+    @classmethod
+    def access(cls, identifier):
+        related_file_obj = (
+            db.session.query(RelatedFile)
+            .filter(RelatedFile.sha256 == identifier)
+            .first()
+        )
+        if related_file_obj is None:
+            return None
+
+        main_obj = (
+            db.session.query(Object)
+            .filter(Object.id == related_file_obj.object_id)
+            .first()
+        )
+        if not main_obj.has_explicit_access(g.auth_user):
+            return None
+
+        return related_file_obj
```

Comment:
As in previous comments, I would just try to perform `Object.access` on parent object and check if related file is actually related with that object_id.

I also see that in current model the same sha256 can't be related with multiple objects? We can actually have two approaches here:
1) Make many-to-many relationship between Object and RelatedFile
2) Consider unique RelatedFile.sha256 only within the same Object.id (but nevertheless we need to choose that Object.sha256 in endpoint parameters)

Reply:
'Object.access()' is much more complex function i.e. it checks object's parents and it uses 1 or 2 queries and outerjoin
I think my solutions is more efficient - I achieve the same result with one query and in this case it should be enough

Reply:
I've decided to implement the second approach
Relation 'Object <-> Related' file is now one to many

I know that one related file can be associated with many objects, but I think that it will happen very rarely.
In my opinion, implementing would be an overkill